sigstore/sign.py

```diff
@@ -188,3 +212,62 @@ class SigningResult(BaseModel):
     """
     A record of the Rekor log entry for the signing operation.
     """
+
+    def _to_bundle(self) -> Bundle:
+        """
+        Creates a Sigstore bundle (as defined by Sigstore's protobuf specs)
+        from this `SigningResult`.
+        """
+
+        # TODO: Should we include our Fulcio intermediate in the chain?
+        chain = X509CertificateChain(
+            certificates=[X509Certificate(raw_bytes=self.cert_pem.encode())]
+        )
```

Comment:
Flagging for review: I wasn't sure about this either. During discussions in protobuf-specs it sounds like the root cert isn't expected to be included, but that clients *may* include any intermediates in use.

Reply:
`sigstore-js` doesn't seem to have problems verifying even without us embedding the intermediate, so I think this is probably fine for now.

Reply:
I think you're supposed to include it based on the docs?

https://github.com/sigstore/protobuf-specs/blob/16541696de137c6281d66d075a4924d9bbd181ff/protos/sigstore_common.proto#L170-L173

CC @haydentherapper  who will have a strong opinion here

Reply:
Hmm, my reading of the language there:

```
        // The first certificate in the array must be the leaf
        // certificate used for signing. Any intermediate certificates
        // must be stored as offset 1 to n-1, and the root certificate at
        // position n.
```

Is that the certificate list MUST be nonempty, the first member MUST be the leaf, and *any if present* intermediates and root MUST be listed in chain order. In other words, just the leaf is fine, but @haydentherapper can confirm!

Reply:
Yeah, we could stand to have an additional clarifying comment for `verification_material`.

Reply:
The intent was that the leaf certificate MUST be included. Any intermediate (if existing), the and the root are RECOMMENDED to be included. 

Regardless if they are included or not in the bundle. The intermediates and the root must of course be verified against the trust root, as this is the only way to securely verify that they are trusted. The reason for not only depend on the root certificate, but also the intermediate is that we don't ship any CRLs, if an intermediate is revoed, it's removed from the TUF root, and so a client MUST NOT trust it.

Reply:
But agree, lets clarify this. The validation part is covered by the client spec, but I think the language is a bit unclear there too.

Reply:
Historical context from Web PKI: Root CAs that follow CA/B forum requirements (typically those that are a part of OS or Mozilla trust stores) have their signing keys offline, and have long lifetimes (10-20 years). Intermediate CAs are created with online signing keys and shorter lifetimes (~3 years) to issue end-entity (leaf) certificates. Because these signing keys are online, there's a higher chance of compromise. In the event of a compromise, a root CA maintains a CRL containing intermediate CA certs that clients are expected to check for intermediate CA trustworthiness.

TLS connections provide the full cert chain, because intermediates are meant only for chain building, to find a path from an end-entity certificate to a root. The root is provided I assume just for auditing purposes, but a verifier should never trust a root provided in a chain, it should come from a trust store. The intermediates build a chain from the leaf to the root, with each intermediate's status being checked against a CRL.

Back to the present: The Sigstore public good instance's root CA is online, because maintaining offline CAs is a pain, especially with this model of cross-company production ownership. We could have skipped using an intermediate and issued end-entity certs directly off of the root, but we chose to leverage an intermediate because it allowed us to convert the Sigstore root CA into a PKI root for all use-cases across Sigstore (currently just Fulcio, but it could include a timestamp authority too).

We didn't want to maintain CRLs because clients struggle with checking them. Either you need to always query an CRL online (no more offline verification then), or cache CRLs and pick some cadence to update. Size can also become an issue, but not worried about that here since root CA CRLs rarely grow large.

But we don't need CRLs for the public good instance, because we have TUF! We put the intermediate and the root in TUF, and so if either become untrusted, we simply remove them from the TUF trust root. Clients (Cosign for example) fetch the verification material from TUF, and for OCI, ignore any certs set in the `chain` annotation (if the TUF root does not contain intermediates, then Cosign does read only the intermediates, not the root, from the chain annotation)

**tl;dr - For cases of verification from the public good instance, it is sufficient to include only the leaf/end-entity certificate in the bundle**

But wait! But what about non-public good instance Sigstore, private Sigstore instances? If they aren't using TUF, this is an issue if they want to rely on a more standard-pki-like experience. Maybe they don't want to ship intermediates in their trust root and rely on CRLs (there was an issue in Cosign about using CRLs). For these use-cases, it would be preferable to include the full chain in the bundle (the root is never necessary, but like I said above, I think it's useful for auditing purposes).

So **second tl;dr - I would consider including the full chain, but if you don't want to support this now, I would at least document this as a limitation/isssue**

Reply:
Thanks for the extremely detailed explanation @haydentherapper! 

Including the intermediate and root certs would require some modifications to the current `SigningResult` model, but not impractical ones. Given that, I'll add a comment here and do that in a follow-up (since I plan on more generally refactoring `SigningResult` per https://github.com/sigstore/sigstore-python/pull/465#discussion_r1082935187 anyways).

Reply:
56f0636..8651f95